### PR TITLE
[url_launcher] Add support for setting show title on Chrome Custom Tabs

### PR DESCRIPTION
**Pull request by Alex-Usmanov**
- Added showTitle option to platform interface & app-facing api
- Made use of that parameter on android platform in order to set show title on Chrome Custom Tabs, when it is being used
- Made android implementation use both legacy & new api (instead of passthrough from new to legacy), in order to allow use of the aforementioned parameter
- Added a button to the example to test this parameter
- Added tests

Screenshot_20231018-095928
(Notice title at the top of the page "CyLog Software - HTTP ...")

https://github.com/flutter/flutter/issues/136784

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/url_launcher/url_launcher_android/lib/url_launcher_android.dart [hunk 1]
```diff
@@ -62,17 +62,42 @@ class UrlLauncherAndroid extends UrlLauncherPlatform {
     required Map<String, String> headers,
     String? webOnlyWindowName,
   }) async {
+    final WebViewOptions webViewOptions = WebViewOptions(
```
Comment (Alex-Usmanov) [inline packages/url_launcher/url_launcher_android/lib/url_launcher_android.dart:65]: There is a fair share of code duplication here, but it could only be resolved after https://github.com/flutter/flutter/issues/66721 unfortunately (could not make it a passthrough because there is no way to pass the useWebView parameter to launchUrl)

**Comment by ditman** [comment]
This cannot be implemented on the web, since the browser controls how tabs/windows are displayed.

File: packages/url_launcher/url_launcher_macos/example/pubspec.yaml
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_macos/example/pubspec.yaml:1]: All the packages where the only change is `dependency_overrides` can be reverted.

File: packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart [hunk 2]
```diff
@@ -44,6 +45,11 @@ class InAppWebViewConfiguration {
 
   /// Additional headers to pass in the load request.
   final Map<String, String> headers;
+
+  /// Whether or not to show the webpage title when using Chrome Custom Tabs.
+  ///
+  /// Only works on Android.
+  final bool showTitle;
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart:52]: Adding this to `InAppWebViewConfiguration` becomes somewhat problematic in light of https://github.com/flutter/packages/pull/5155.

We could potentially overload this to apply to browser views as well, but given that the set of things that work for web views and the set of things that work for browser views is likely to stay completely disjoint, it would probably be better to add a new configuration object for browser view settings instead. (It won't just be this one; SFSafariViewController has a few options, for instance.)

**Comment by Alex-Usmanov** [comment]
- Added a configuration object for in app browser
- Removed unnecessary dependency overrides

*Will update docs once #5155 lands

File: packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart (hunk 2 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart:62]: Good point! Added `InAppBrowserConfiguration` to address this

File: packages/url_launcher/url_launcher_macos/example/pubspec.yaml
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_macos/example/pubspec.yaml:1]: Did not think of that, done ✅

File: packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart:1]: This package will need a version bump and changelog.

File: packages/url_launcher/url_launcher_android/test/url_launcher_android_test.dart [hunk 3]
```diff
@@ -415,7 +431,15 @@ class _FakeUrlLauncherApi implements UrlLauncherApi {
   @override
   Future<bool> launchUrl(String url, Map<String?, String?> headers) async {
     passedWebViewOptions = WebViewOptions(
-        enableJavaScript: false, enableDomStorage: false, headers: headers);
+      enableJavaScript: false,
+      enableDomStorage: false,
+      headers: headers,
+    );
+
+    passedBrowserOptions = BrowserOptions(
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/test/url_launcher_android_test.dart:439]: Why is this being set?

File: packages/url_launcher/url_launcher_android/pigeons/messages.dart [hunk 4]
```diff
@@ -25,6 +27,12 @@ class WebViewOptions {
   final Map<String?, String?> headers;
 }
 
+class BrowserOptions {
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/pigeons/messages.dart:30]: The class and property should have doc comments.

File: packages/url_launcher/url_launcher_android/pigeons/messages.dart [hunk 5]
```diff
@@ -33,9 +41,14 @@ abstract class UrlLauncherApi {
   /// Opens the URL externally, returning true if successful.
   bool launchUrl(String url, Map<String, String> headers);
 
-  /// Opens the URL in an in-app Custom Tab or WebView, returning true if it
-  /// opens successfully.
-  bool openUrlInApp(String url, bool allowCustomTab, WebViewOptions options);
+  /// Opens the URL in an in-app WebView, returning true if it opens
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/pigeons/messages.dart:44]: Why was "Custom Tab" removed from this comment?

File: packages/url_launcher/url_launcher_android/example/integration_test/url_launcher_test.dart
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/example/integration_test/url_launcher_test.dart:1]: This file's change can be reverted.

File: packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java [hunk 6]
```diff
@@ -140,10 +144,16 @@ public void closeWebView() {
     return CustomTabsClient.getPackageName(applicationContext, Collections.emptyList()) != null;
   }
 
-  private static boolean openCustomTab(
-      @NonNull Context context, @NonNull Uri uri, @NonNull Bundle headersBundle) {
-    CustomTabsIntent customTabsIntent = new CustomTabsIntent.Builder().build();
+  @VisibleForTesting
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java:147]: Why was this made public for testing? Testing should be done via public interface whenever possible to make tests easier to maintain and less likely to miss important behavior.

File: packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java [hunk 7]
```diff
@@ -169,7 +179,8 @@ private static boolean containsRestrictedHeader(Map<String, String> headersMap)
     return false;
   }
 
-  private static @NonNull Bundle extractBundle(Map<String, String> headersMap) {
+  @VisibleForTesting
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java:182]: Same question here.

File: packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java [hunk 8]
```diff
@@ -239,6 +245,68 @@ public void openWebView_opensUrlInCustomTabsWithCORSAllowedHeader() {
...
 
+  @Test
+  public void openWebView_opensUrlInCustomTabsWithShowTitle() {
+    Activity activity = mock(Activity.class);
+    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
+    api.setActivity(activity);
+    String url = "https://flutter.dev";
+    HashMap<String, String> headers = new HashMap<>();
+    String headerKey = "Content-Type";
+    headers.put(headerKey, "text/plain");
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java:256]: Headers don't seem relevant to this test, so shouldn't need to be set up.

File: packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java [hunk 9]
```diff
@@ -239,6 +245,68 @@ public void openWebView_opensUrlInCustomTabsWithCORSAllowedHeader() {
...
+
+  @Test
+  public void openWebView_opensUrlInCustomTabsWithoutShowTitle() {
+    Activity activity = mock(Activity.class);
+    UrlLauncher api = new UrlLauncher(ApplicationProvider.getApplicationContext());
+    api.setActivity(activity);
+    String url = "https://flutter.dev";
+    HashMap<String, String> headers = new HashMap<>();
+    String headerKey = "Content-Type";
+    headers.put(headerKey, "text/plain");
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java:287]: Same.

File: packages/url_launcher/url_launcher/lib/src/types.dart [hunk 10]
```diff
@@ -55,3 +55,15 @@ class WebViewConfiguration {
...
+
+/// Additional configuration options for [LaunchMode.inAppBrowserView]
+@immutable
+class BrowserConfiguration {
+  /// Creates a new InAppBrowserConfiguration with given settings.
+  const BrowserConfiguration({this.showTitle = false});
+
+  /// Whether or not to show the webpage title when using Chrome Custom Tabs.
+  ///
+  /// Only works on Android.
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher/lib/src/types.dart:67]: Same comments here as on the platform interface; platform details shouldn't be documented at this level.

File: packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart [hunk 11]
```diff
@@ -50,13 +50,26 @@ class InAppWebViewConfiguration {
   final Map<String, String> headers;
 }
 
+/// Additional configuration options for [PreferredLaunchMode.inAppBrowserView]
+@immutable
+class InAppBrowserConfiguration {
+  /// Creates a new InAppBrowserConfiguration with given settings.
+  const InAppBrowserConfiguration({this.showTitle = false});
+
+  /// Whether or not to show the webpage title when using Chrome Custom Tabs.
```
[the remark this answers is not in the document]
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart]: Done

File: packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart [hunk 12]
```diff
@@ -50,13 +50,26 @@ class InAppWebViewConfiguration {
...
+  ///
+  /// Only works on Android.
```
[the remark this answers is not in the document]
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart]: Done

File: packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart
[the remark this answers is not in the document]
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart:1]: Done. Did not quite think of a lot of testcases, you can find those in packages/url_launcher/url_launcher_platform_interface/test/in_app_browser_configuration_test.dart and packages/url_launcher/url_launcher_platform_interface/test/launch_options_test.dart

Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart:1]: Done

File: packages/url_launcher/url_launcher_android/pubspec.yaml
[the remark this answers is not in the document]
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/pubspec.yaml:1]: Done

File: packages/url_launcher/url_launcher_android/test/url_launcher_android_test.dart (hunk 3 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/test/url_launcher_android_test.dart]: You are right, this doesn't make sense. Removed that

File: packages/url_launcher/url_launcher_android/pigeons/messages.dart (hunk 4 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/pigeons/messages.dart:31]: Added those

File: packages/url_launcher/url_launcher_android/pigeons/messages.dart (hunk 5 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/pigeons/messages.dart]: Cannot quite recall. Looking back I had no reason to do that. Reverted

File: packages/url_launcher/url_launcher_android/example/integration_test/url_launcher_test.dart
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/example/integration_test/url_launcher_test.dart:1]: Done

File: packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java (hunk 6 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java]: Reverted this change as well

File: packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java (hunk 7 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/android/src/main/java/io/flutter/plugins/urllauncher/UrlLauncher.java]: Done

File: packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java (hunk 8 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java]: Yes, you are right. Removed this

File: packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java (hunk 9 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher_android/android/src/test/java/io/flutter/plugins/urllauncher/UrlLauncherTest.java]: Done

File: packages/url_launcher/url_launcher/lib/src/types.dart (hunk 10 above)
Comment (Alex-Usmanov) [reply packages/url_launcher/url_launcher/lib/src/types.dart]: Made this comment platform-independent

**Comment by Alex-Usmanov** [comment]
Thank you for your valuable feedback and your time! 
I am pretty sure I've fixed everything regarding the code part of this PR. CI seems to be failing due to specifics of landing PRs. I can split this up in two as soon as we get over the final touches

File: packages/url_launcher/url_launcher/example/lib/main.dart [hunk 13]
```diff
@@ -220,6 +230,13 @@ class _MyHomePageState extends State<MyHomePage> {
                 child: const Text('Launch in app + close after 5 seconds'),
               ),
               const Padding(padding: EdgeInsets.all(16.0)),
+              ElevatedButton(
+                onPressed: () => setState(() {
+                  _launched = _launchInAppWithBrowserOptions(toLaunch);
+                }),
+                child: const Text('Launch in app with showTitle'),
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher/example/lib/main.dart:237]: "[...] with title displayed"

File: packages/url_launcher/url_launcher_android/CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,7 @@
-## NEXT
+## 6.3.0
 
+* Implement taking in `BrowserConfiguration` parameter
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/CHANGELOG.md:3]: These need to follow the CHANGELOG style guide linked from the PR checklist (verb form, punctuation).

File: packages/url_launcher/url_launcher_android/pigeons/messages.dart [hunk 15]
```diff
@@ -25,6 +27,16 @@ class WebViewOptions {
   final Map<String?, String?> headers;
 }
 
+/// Configuration options for in-app browser views.
+class BrowserOptions {
+  BrowserOptions({required this.showTitle});
+
+  /// Whether or not to show the webpage title.
+  ///
+  /// May not be supported on all platforms.
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_android/pigeons/messages.dart:36]: This file is internal to the Android implementation, so this sentence doesn't make sense here.

File: packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart [hunk 16]
```diff
@@ -50,13 +50,26 @@ class InAppWebViewConfiguration {
   final Map<String, String> headers;
 }
 
+/// Additional configuration options for [PreferredLaunchMode.inAppBrowserView]
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_platform_interface/lib/src/types.dart:53]: Nit: Missing a period.

**Review by stuartmorgan — COMMENTED** [review]
This looks good with a few minor nits; please feel free to split out the first sub-PR with just the platform interface changes.

**Comment by stuartmorgan** [comment]
Update from triage: This is blocked on https://github.com/flutter/packages/pull/5759, which is currently in review.

**Comment by Alex-Usmanov** [comment]
I made a terrible mistake with source control and https://github.com/flutter/packages/pull/5166 was automatically closed. The restored version is in #6097. 

Once again sorry for the inconvenience 😔
